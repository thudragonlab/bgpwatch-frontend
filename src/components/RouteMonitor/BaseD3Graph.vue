<template>
  <div class="D3_graph-box" :id="d3Id"></div>
</template>

<script>
import * as d3 from 'd3'
import * as dagreD3 from 'dagre-d3'
export default {
  name: 'D3Graph',
  created() {},
  data() {
    return {}
  },
  props: {
    d3Id: {
      type: String,
      defalut: 'graph',
    },
    links: {
      type: Array,
      defalut: [],
    },
    nodes: {
      type: Array,
      defalut: [],
    },
    customerFunc: {
      type: Function,
      defalut: () => {},
    },
    moreAction: {
      type: Function,
      defalut: () => {},
    },
    options: {
      type: Object,
      default: () => {
        return {
          rankdir: 'LR',
          ranksep: 40, // 节点之间线距离
          nodesep: 10, // 节点之间距离
          marginx: 10,
        }
      },
    },
  },
  watch: {
    nodes(newV) {
      this.drawGraph(this.links, this.nodes, this.d3Id)
    },
  },
  mounted() {
    if (this.links && this.links.length !== 0 && this.nodes && this.nodes.length !== 0) {
      this.drawGraph(this.links, this.nodes, this.d3Id)
    }
  },
  methods: {
    drawGraph(linkArray, nodes, svgId) {
      // 获取所有节点
      const g = this.createInstanceofGraph()
      this.setNode(g, nodes)
      this.setEdge(g, linkArray)
      this.customerFunc(g, linkArray, nodes)
      this.renderGraph(g, svgId)
    },
    createInstanceofGraph() {
      return new dagreD3.graphlib.Graph().setGraph(this.options).setDefaultEdgeLabel(function () {
        return {}
      })
    },
    setNode(g, nodes) {
      var n_nodes = nodes.length
      for (var i = 0; i < n_nodes; i++) {
        g.setNode(nodes[i], {
          label: nodes[i],
          labelStyle: 'font-size: 10px',
          shape: 'circle',
          style: 'stroke: darkgray; fill: Gainsboro',
        })
      }
    },
    setEdge(g, linkArray) {
      var n_edges = linkArray.length
      for (var i = 0; i < n_edges; i++) {
        var edge = linkArray[i]
        g.setEdge(edge['from'], edge['to'], {
          style: 'stroke: rgb(20, 20, 20); fill: none;',
          arrowheadStyle: 'fill: rgb(20, 20, 20); stroke: rgb(20, 20, 20);',
          arrowhead: 'vee',
        })
      }
    },
    renderGraph(g, svgId) {
      var render = new dagreD3.render()
      var svg1 = d3.select(`#${svgId}`)
      if (svg1.selectAll('svg')._groups[0]?.length !== 0) {
        svg1.select('svg').remove() // remove previous nodes, clear the canvas
      }
      var svg = svg1.append('svg')
      var inner = svg.append('g')
      render(inner, g)
      this.moreAction(g, inner, svg, `${svgId}`)
    },
  },
}
</script>

<style lang="scss" scoped>
.D3_graph-box {
  width: 100%;
  // background: #000;
}
</style>
