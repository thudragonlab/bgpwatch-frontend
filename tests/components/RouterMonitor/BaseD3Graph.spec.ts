import {
    Wrapper,
    mount,
    shallowMount
  } from '@vue/test-utils'
  import BaseD3Graph from "@/components/RouteMonitor/BaseD3Graph.vue";
  import ElementUI from '@/lib/element.js'
  import Vue from "vue";
  
  import 'jest-canvas-mock'
  
  Vue.use(ElementUI)
  
  
  describe('BaseD3Graph.vue', () => {
         // 测试节点和边的信息是否正确传递并渲染成功
  it('renders nodes and links correctly', async () => {
    const links = [
      { from: 'node1', to: 'node2' },
      { from: 'node2', to: 'node3' }
    ]
    const nodes = ['node1', 'node2', 'node3']
    const wrapper:Wrapper<any> = mount(BaseD3Graph, {
      propsData: {
        d3Id:'ggg',
        links,
        nodes,
        customerFunc:jest.fn(),
        moreAction:jest.fn()
      }
    })
    expect(wrapper.html()).toMatchSnapshot();
    const drawGraph = jest.fn()
    wrapper.vm.drawGraph = drawGraph
    await wrapper.setProps({
        nodes:['node1', 'node2']
    })
    expect(drawGraph).toHaveBeenCalled()

  })
  // 测试自定义操作函数是否被正确调用
  it('calls custom function correctly', () => {
    const links = [
      { from: 'node1', to: 'node2' }
    ]
    const nodes = ['node1', 'node2']
    const customerFunc = jest.fn((graph, links, nodes) => {
      graph.setNode('node3', {
        label: 'node3',
        labelStyle: 'font-size: 10px',
        shape: 'circle',
        style: 'stroke: darkgray; fill: Gainsboro',
      })
      graph.setEdge('node2', 'node3', {
        style: 'stroke: rgb(20, 20, 20); fill: none;',
        arrowheadStyle: 'fill: rgb(20, 20, 20); stroke: rgb(20, 20, 20);',
        arrowhead: 'vee',
      })
    })
    const wrapper = mount(BaseD3Graph, {
      propsData: {
        links,
        nodes,
        customerFunc,
        moreAction:jest.fn()
      }
    })
    expect(customerFunc).toHaveBeenCalledWith(expect.any(Object), links, nodes)
  })
  // 测试更多操作函数是否被正确调用
  it('calls more action function correctly', () => {
    const links = [
      { from: 'node1', to: 'node2' }
    ]
    const nodes = ['node1', 'node2']
    const moreAction = jest.fn((graph, inner, svg, svgId) => {
      svg.attr('width', '100%')
      svg.attr('height', '100%')
    })
    const wrapper = mount(BaseD3Graph, {
      propsData: {
        links,
        nodes,
        customerFunc:jest.fn(),
        moreAction

      }
    })
    expect(moreAction).toHaveBeenCalledWith(expect.any(Object), expect.any(Object), expect.any(Object), expect.any(String))
  })
    
  })