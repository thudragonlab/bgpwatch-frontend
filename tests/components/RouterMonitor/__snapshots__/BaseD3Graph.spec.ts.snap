// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseD3Graph.vue renders nodes and links correctly 1`] = `<div id="ggg" class="D3_graph-box"></div>`;
